server/shared/database/migrations/20200215090218-add_modified-at-to-activity.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+const filter = require('lodash/filter');
+const get = require('lodash/get');
+const head = require('lodash/head');
+const keyBy = require('lodash/keyBy');
+const map = require('lodash/map');
+const max = require('lodash/max');
+const Promise = require('bluebird');
+const transform = require('lodash/transform');
+
+const TABLE = 'activity';
+const COLUMN = 'modified_at';
+
+const isOutline = it => /\//.test(it.type);
+
+exports.up = async (qi, { DATE }) => {
+  await qi.addColumn(TABLE, COLUMN, { type: DATE });
+  await updateModifiedAt(qi);
+};
+
+exports.down = qi => qi.removeColumn(TABLE, COLUMN);
+
+async function updateModifiedAt({ sequelize }) {
+  const transaction = await sequelize.transaction();
+  const repositories = await getRepositories(sequelize, transaction);
+  await Promise.map(repositories, async ({ id: repositoryId }) => {
+    const [activities, revisions] = await Promise.all([
+      getActivities(sequelize, repositoryId, transaction),
+      getRevisions(sequelize, repositoryId, transaction)
+    ]);
+    const cache = buildRevisionCache(activities, revisions);
+    const outlineActivities = filter(activities, isOutline);
+    await Promise.map(outlineActivities, activity => {
+      const modifiedAt = resolveModifiedAt(activity, cache);
+      return setModifiedAt(sequelize, transaction, activity, modifiedAt);
+    });
+  });
+  await transaction.commit();
+}
+
+async function getRepositories(sequelize, transaction) {
+  const sql = 'SELECT id AS "id" FROM repository;';
+  return head(await sequelize.query(sql, { transaction, raw: true }));
+}
+
+async function getRevisions(sequelize, repositoryId, transaction) {
+  const sql = `
+    SELECT
+      entity AS "entity",
+      state -> 'id' AS "entityId",
+      state -> 'parentId' AS "parentId",
+      repository_id AS "repositoryId",
+      CASE
+        WHEN
+          entity = 'ACTIVITY'
+        THEN
+          state -> 'id'
+        ELSE
+          state -> 'activityId'
+      END
+        AS "activityId",
+      MAX(created_at) AS "createdAt"
+    FROM
+      revision
+    WHERE
+      repository_id = :repositoryId AND
+      entity IN ( 'ACTIVITY', 'CONTENT_ELEMENT' )
+    GROUP BY
+      "entity", "entityId", "parentId", "repositoryId", "activityId";
+  `;
+  const options = { transaction, replacements: { repositoryId }, raw: true };
+  return head(await sequelize.query(sql, options));
+}
+
+async function getActivities(sequelize, repositoryId, transaction) {
+  const sql = `
+    SELECT
+      id AS "id",
+      type AS "type",
+      parent_id AS "parentId",
+      repository_id AS "repositoryId"
+    FROM
+      activity
+    WHERE
+      repository_id = :repositoryId;
+  `;
+  const options = { transaction, replacements: { repositoryId }, raw: true };
+  return head(await sequelize.query(sql, options));
+}
+
+function setModifiedAt(sequelize, transaction, activity, modifiedAt) {
```

Comment:
Move transaction as last parameter. Group activityId and modifiedAt within `where` parameter (use for replacements).